### PR TITLE
[FLINK-17596][python] Move the Python UDF runner script out of the jar of flink-python

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request refactors the ProcessPythonEnvironmentManager to move the pyflink udf runner script to the pyflink python package.*

## Brief change log

  - *Refactor the ProcessPythonEnvironmentManager to prepare the python environment from Java side.*
  - *Move the pyflink udf runner script to the pyflink python package.*

## Verifying this change

This change is already covered by existing tests, such as *test_dependency.py*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		}
+		if (!new File(runnerScriptPath).exists()) {
+			throw new FileNotFoundException(String.format(
+				"The runner script '%s' does not exist! " +
+				"Please reinstall the apache-flink Python package.", runnerScriptPath));
+		}
+		return runnerScriptPath;
+	}
+
+	private static String getSitePackagesPath(
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java:133]: Why need to add the site package path to `PYTHONPATH` manually?

File: flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+		"len(installed_dir) > 1 and " +
+			"print(installed_dir[1])";
+
+	private static final String CHECK_PIP_VERSION_SCRIPT =
+		"import sys;" +
+		"from pkg_resources import get_distribution, parse_version;" +
+		"pip_version = get_distribution('pip').version;" +
+		"print(parse_version(pip_version) >= parse_version(sys.argv[1]))";
+
+	private static final String GET_RUNNER_SCRIPT_DIR =
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java:70]: The constant name should be consistent, e.g. if we add the suffix `_SCRIPT`, we should make sure that all the constants have this suffix.

File: flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+		"import pyflink;" +
+		"import os;" +
+		"print(os.path.join(os.path.abspath(os.path.dirname(pyflink.__file__)), 'bin'))";
+
+	public static void pipInstallRequirements(
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java:75]: Add some Javadoc for the util methods?

File: flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Utils used to install the python requirements via pip.
+ */
+public class PythonHelperScriptUtils {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java:42]: What about rename to `PythonEnvironmentManagerUtils`?

File: flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+			@Nullable String requirementsCacheDir,
+			String requirementsInstallDir,
+			String pythonExecutable,
+			Map<String, String> environmentVariables) throws IOException, InterruptedException {
+		String sitePackagesPath = getSitePackagesPath(requirementsInstallDir, pythonExecutable, environmentVariables);
+		String path = String.join(File.pathSeparator, requirementsInstallDir, "bin");
+		appendToEnvironmentVariable("PYTHONPATH", sitePackagesPath, environmentVariables);
+		appendToEnvironmentVariable("PATH", path, environmentVariables);
+
+		List<String> commands = new ArrayList<>(Arrays.asList(
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java]: The `Arrays.asList` returns a immutable List object but we need a mutable List, so we need to copy the content to the `ArrayList`.

File: flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+		pb.redirectErrorStream(true);
+		Process p = pb.start();
+		InputStream in = new BufferedInputStream(p.getInputStream());
+		BufferedReader br = new BufferedReader(new InputStreamReader(in));
+		StringBuilder out = new StringBuilder();
+		String s;
+		while ((s = br.readLine()) != null) {
+			out.append(s).append("\n");
+		}
+		if (p.waitFor() != 0) {
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java]: I think it is necessary to use `waitFor(long timeout, TimeUnit unit)` as the subprocess is already exited when the `br.readLine` return `null`. I agree that the InterruptedException could be ignored because the `waitFor` method would not be blocked in this situation.

File: flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java (hunk 1 above)
Comment (WeiZhong94) [reply flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java]: Because this "site-packages" is a temporary directory, which would not be placed to the `PYTHONPATH` by default.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have addressed your comments.

File: flink-python/pyflink/table/tests/test_dependency.py [hunk 7]
```diff
@@ -130,9 +129,22 @@ def test_set_requirements_with_cached_directory(self):
         os.mkdir(requirements_dir_path)
         package_file_name = "python-package1-0.0.0.tar.gz"
         with open(os.path.join(requirements_dir_path, package_file_name), 'wb') as f:
-            from pyflink.fn_execution.tests.process_mode_test_data import file_data
             import base64
-            f.write(base64.b64decode(json.loads(file_data[package_file_name])["data"]))
+            # This base64 data is encoded from a python package file which includes a
+            # "python_package1" module. The module contains a "plus(a, b)" function.
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_dependency.py:134]: could you add some description on how to recompute the base64?

File: flink-python/src/main/java/org/apache/flink/python/env/ProcessPythonEnvironmentManager.java [hunk 8]
```diff
@@ -221,7 +235,7 @@ public String createRetrievalToken() throws IOException {
 		// disable the launching of gateway server to prevent from this dead loop:
 		// launch UDF worker -> import udf -> import job code
 		//        ^                                    | (If the job code is not enclosed in a
-		//        									   |  if name == 'main' statement)
+		//        |								       |  if name == 'main' statement)
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/env/ProcessPythonEnvironmentManager.java:238]: revert this change?

File: flink-python/src/main/java/org/apache/flink/python/util/PythonEnvironmentManagerUtils.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+			boolean log) throws IOException {
+		ProcessBuilder pb = new ProcessBuilder(commands);
+		pb.environment().putAll(environmentVariables);
+		pb.redirectErrorStream(true);
+		Process p = pb.start();
+		InputStream in = new BufferedInputStream(p.getInputStream());
+		StringBuilder out = new StringBuilder();
+		String s;
+		if (log) {
+			LOG.info(String.format("Execute the commnad: %s", String.join(" ", commands)));
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonEnvironmentManagerUtils.java:178]: typo: command

Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonEnvironmentManagerUtils.java:178]: `Executing command: %s`

File: flink-python/src/main/java/org/apache/flink/python/util/PythonEnvironmentManagerUtils.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+	private static void appendToEnvironmentVariable(String key, String path, Map<String, String> env) {
+		if (env.containsKey(key)) {
+			env.put(key, String.join(File.pathSeparator, path, env.get(key)));
+		} else {
+			env.put(key, path);
+		}
+	}
+
+	public static void main(String[] args) throws IOException, InterruptedException {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonEnvironmentManagerUtils.java:208]: remove this?

File: flink-python/src/main/java/org/apache/flink/python/util/PythonEnvironmentManagerUtils.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+			throw new FileNotFoundException(String.format(
+				"The runner script '%s' does not exist! " +
+				"Please reinstall the apache-flink Python package.", runnerScriptPath));
+		}
+		return runnerScriptPath;
+	}
+
+	private static String getSitePackagesPath(
+			String prefix,
+			String pythonExecutable,
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonEnvironmentManagerUtils.java:150]: it will not throw `InterruptedException` any more. Should also check the other places.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have addressed your comments and rebased the commit on the top of the master branch.

**Comment by WeiZhong94** [comment]
@dianfu The test has passed in my azure. https://dev.azure.com/weizhong0618/Flink/_build/results?buildId=101&view=results

**Comment by dianfu** [comment]
Thanks. Merging...
